docs/sdg/sdg-api-interface.md

```diff
@@ -0,0 +1,71 @@
+# SDG Library Design
+
+
+## Objective
+Library called `instructlab-sdg` that can be called per seed example that includes question and answer pairs, and context for grounded skills.
+
+## Structure of the SDG Library
+
+We propose the following structure for the SDG library. There will be config files that contain all the prompt templates for the pipelines. 
+
+
+```
+- src/instructlab/sdg/
+    - configs/ 
+        - gen_q.yaml 
+        - gen_a.yaml
+        - ...
+    - init.py 
+    - block.py 
+    - llmblock.py 
+    - pipeline.py 
+    - sdg.py
+```
+
+
+## CLI
+
+The CLI client uses the instructlab SDG library and provides it a run configuration with input parameters.
+
+
+```
+# cli_driver.py
+
+import sdg
+
+client = openai_client(endpoint)
+cli_sdg = sdg(run_config, client)  # run config has all the variables like num_samples, pipelinesteps etc
+generated_samples = cli_sdg.generate()
+```
+
+The run configuration contains the parameters needed to run the SDG code library as well as points to the templates needed to run the SDG code as well as the prompt template and the default model system prompt template.
+
+* num_samples is the number of synthetic samples that you wish to generate per seed example
+* max_retry is the maximum number of non-greedy retries you want to make if the num_samples is not reached. The number of samples in the generated output will be the samples achieved until max_retry is reached
+* Pipeline steps contains the steps that you want to invoke in the SDG pipeline and the prompt configurations per step. The variable names of the blocks can be anything and the prompt configs must be compatible with the teacher model
+*  max_new_tokens is the maximum number of tokens we want to generate. In other words, the size of the output sequence, not including the tokens in the prompt. 
+*  model_name is the teacher model we would want to use to generate the synthetic data.
+*  model_template and stop_token are the paramaters for generation template.
```

Comment:
Some suggestions for config items based on what the CLI uses today:

https://github.com/instructlab/sdg/blob/7e3a83e3385967828875e95ae5e6f11123e2ab35/src/instructlab/sdg/generate_data.py#L365-L390

- `output_dir` -- where to put the generated samples
- `num_cpus` -- how much to parallelize our inference requests.
- `server_ctx_size` / `chunk_word_count` -- do we need any options to inform how to chunk large knowledge docs?

There's others that I'm not sure we even use ... it's worth a look to see if you spot anything you think should be in the config

Reply:
* SDG will return generated objects in-memory, and the cli driver will configure where to save the data, and we can have an `output_dir` param to help configure that.
* We can pass `num_cpus` as a param to SDG
* `chunk_word_count` or something similar might be parameter that will pass to the SDG library. @PalmPalm7 @suppathak are looking into chunking.